src/Middleware/RequestThrottling/src/QueuePolicies/StackQueuePolicy.cs

```diff
@@ -0,0 +1,104 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.RequestThrottling.Policies
+{
+    internal class StackQueuePolicy : IQueuePolicy
+    {
+        private readonly List<TaskCompletionSource<bool>> _buffer;
+        private readonly int _maxQueueCapacity;
+        private readonly int _maxConcurrentRequests;
+        private bool _hasReachedCapacity;
+        private int _head;
+        private int _queueLength;
+
+        private static readonly Task<bool> _trueTask = Task.FromResult(true);
+
+        private readonly object _bufferLock = new Object();
+
+        private int _freeServerSpots;
+
+        public StackQueuePolicy(IOptions<QueuePolicyOptions> options)
+        {
+            _buffer = new List<TaskCompletionSource<bool>>();
+            _maxQueueCapacity = options.Value.RequestQueueLimit;
+            _maxConcurrentRequests = options.Value.MaxConcurrentRequests;
+            _freeServerSpots = options.Value.MaxConcurrentRequests;
+        }
+
+        public Task<bool> TryEnterAsync()
+        {
+
+            lock (_bufferLock)
+            {
+                if (_freeServerSpots > 0)
```

Comment:
Is it a good idea a this part is moved back to the middleware, so the `IQueuePolicy` only handles the queue? So if some one wants to implement a custom `IQueuePolicy` does not need to repeat this.

Reply:
I thought about doing this, and in an ideal world would split the logic this way. 

In the other implementation, a single SemaphoreSlim controls both flow into the server, and the queue on top. In this implementation, the queue is separate, and flow is control by this semaphore baked into the queue logic. I could go for more abstraction by moving this logic up to the middleware, but I'm trying to keep policy design open-ended and prevent extra lock congestion.

Reply:
@Kahbazi I like the way you think. [We've discussed this](https://github.com/aspnet/AspNetCore/pull/11167#discussion_r293175216), but we've decided that the extra purity isn't worth extra development effort yet.

One thing I don't like about leaving the concurrent request limit completely up to the middleware is that the IQueuePolicy could be left unaware how many requests the middleware will try to dequeue in the near future. If the middleware is always trying to dequeue every request all the time, that gives more control to the IQueuePolicy, but that forces the IQueuePolicy to manage a concurrent request limit itself.